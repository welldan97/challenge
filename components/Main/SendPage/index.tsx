import { Formik } from 'formik';
import { memo, useCallback, useMemo } from 'react';
import styled from 'styled-components';
import Input from '../../lib/Input';
import { Box, Button, Footer, Header } from '../styles';
import { Transaction } from '../../../lib/Transaction';

// SECTION: Styles

const Form = styled.form`
  margin-top: 29px;
  margin-bottom: 7px;
`;

// !SECTION

// SECTION: Main

interface SendProps {
  onSend: () => void;
}

const SendPage = ({ onSend }: SendProps) => (
  <Box isShrunk>
    <Header>Please fill the form to send Ethereum</Header>
    <Form onSubmit={onSend} id="transaction">
      <Input name="from" placeholder="Your address" label="From" />
      <Input name="to" placeholder="Destination address" label="To" />
      <Input name="amount" placeholder="Ethereum amount" label="Amount" />
    </Form>
    <Footer isSeamless>
      <Button type="submit" autoFocus onClick={onSend} form="transaction">
        Send
      </Button>
    </Footer>
  </Box>
);

// /SECTION
// SECTION: Container

interface Props {
  onSend: (transaction: Transaction) => void;
}

export default memo(({ onSend }: Props) => {
  const transaction = useMemo(
    () => ({
      from: '',
      to: '',
      amount: '',
    }),
    [],
  );

  return (
    <Formik initialValues={transaction} enableReinitialize onSubmit={onSend}>
      {({ handleSubmit }) => <SendPage onSend={handleSubmit} />}
    </Formik>
  );
});
