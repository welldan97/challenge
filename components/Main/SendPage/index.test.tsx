import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SendPage from '.';

// SECTION: Main

describe('SendPage', () => {
  test('render', () => {
    const onSend = () => {};
    const { container } = render(<SendPage onSend={onSend} />);
    expect(container).toMatchSnapshot();
  });

  test('send', async () => {
    const onSend = jest.fn();
    render(<SendPage onSend={onSend} />);

    const from = screen.getByTestId('from');
    await userEvent.type(from, 'TEST_FROM');

    const to = screen.getByTestId('to');
    await userEvent.type(to, 'TEST_TO');

    const amount = screen.getByTestId('amount');
    await userEvent.type(amount, 'TEST_AMOUNT');

    const send = screen.getByTestId('send');

    await userEvent.click(send);

    expect(onSend.mock.calls[0][0]).toEqual({
      from: 'TEST_FROM',
      to: 'TEST_TO',
      amount: 'TEST_AMOUNT',
    });
  });
});
