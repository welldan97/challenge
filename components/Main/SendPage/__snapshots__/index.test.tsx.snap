// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendPage render 1`] = `
<div>
  <div
    class="sc-da1060d0-0 dVsQgh"
  >
    <header
      class="sc-da1060d0-1 eHCwEr"
    >
      Please fill the form to send Ethereum
    </header>
    <form
      class="sc-b81309d0-0 ePlrqb"
      id="transaction"
    >
      <div
        class="sc-6f71e095-0 zKiGe"
      >
        <label
          class="sc-6f71e095-1 dpiAKW"
          for="from"
        >
          From
        </label>
        <input
          class="sc-6f71e095-2 mOreu"
          data-testid="from"
          id="from"
          placeholder="Your address"
          value=""
        />
      </div>
      <div
        class="sc-6f71e095-0 zKiGe"
      >
        <label
          class="sc-6f71e095-1 dpiAKW"
          for="to"
        >
          To
        </label>
        <input
          class="sc-6f71e095-2 mOreu"
          data-testid="to"
          id="to"
          placeholder="Destination address"
          value=""
        />
      </div>
      <div
        class="sc-6f71e095-0 zKiGe"
      >
        <label
          class="sc-6f71e095-1 dpiAKW"
          for="amount"
        >
          Amount
        </label>
        <input
          class="sc-6f71e095-2 mOreu"
          data-testid="amount"
          id="amount"
          placeholder="Ethereum amount"
          value=""
        />
      </div>
    </form>
    <footer
      class="sc-da1060d0-2 MrGpr"
    >
      <button
        class="sc-da1060d0-3 fUffHN"
        data-testid="send"
        form="transaction"
        type="submit"
      >
        Send
      </button>
    </footer>
  </div>
</div>
`;
